src/matrix/verification/SAS/stages/CalculateSASStage.ts

```diff
@@ -0,0 +1,142 @@
+        const theirInfo = `${sas.their.userId}|${sas.their.deviceId}|${sas.their.publicKey}|`;
+        const sasInfo =
+            "MATRIX_KEY_VERIFICATION_SAS|" +
+            (sas.initiatedByMe ? ourInfo + theirInfo : theirInfo + ourInfo) + sas.id;
+        return olmSAS.generate_bytes(sasInfo, bytes);
+    },
+    "curve25519": function (sas: SASUserInfoCollection, olmSAS: Olm.SAS, bytes: number): Uint8Array {
+        const ourInfo = `${sas.our.userId}${sas.our.deviceId}`;
+        const theirInfo = `${sas.their.userId}${sas.their.deviceId}`;
+        const sasInfo =
+            "MATRIX_KEY_VERIFICATION_SAS" +
+            (sas.initiatedByMe ? ourInfo + theirInfo : theirInfo + ourInfo) + sas.id;
+        return olmSAS.generate_bytes(sasInfo, bytes);
+    },
+} as const;
+
+export class CalculateSASStage extends BaseSASVerificationStage {
+    private resolve: () => void;
+    private reject: (error: VerificationCancelledError) => void;
+
+    public emoji: ReturnType<typeof generateEmojiSas>;
+
+    async completeStage() {
+        await this.log.wrap("CalculateSASStage.completeStage", async (log) => {
+            // 1. Check the hash commitment
+            if (this.needsToVerifyHashCommitment && !await this.verifyHashCommitment(log)) {
+                return;
+            }
+            // 2. Calculate the SAS
+            const emojiConfirmationPromise: Promise<void> = new Promise((res, rej) => {
+                this.resolve = res;
+                this.reject = rej;
+            });
+            this.olmSAS.set_their_key(this.theirKey);
+            const sasBytes = this.generateSASBytes();
+            this.emoji = generateEmojiSas(Array.from(sasBytes));
+            this.eventEmitter.emit("EmojiGenerated", this);
+            await emojiConfirmationPromise;
+            this.setNextStage(new SendMacStage(this.options));
+        });
+    }
+
+    async verifyHashCommitment(log: ILogItem) {
+        return await log.wrap("CalculateSASStage.verifyHashCommitment", async () => {
+            const acceptMessage = this.channel.getReceivedMessage(VerificationEventTypes.Accept).content;
+            const keyMessage = this.channel.getReceivedMessage(VerificationEventTypes.Key).content;
+            const commitmentStr = keyMessage.key + anotherjson.stringify(this.channel.startMessage.content);
+            const receivedCommitment = acceptMessage.commitment;
+            const hash = this.olmUtil.sha256(commitmentStr);
+            if (hash !== receivedCommitment) {
+                log.log({l: "Commitment mismatched!", received: receivedCommitment, calculated: hash});
+                await this.channel.cancelVerification(CancelTypes.MismatchedCommitment);
+                return false;
+            }
+            return true;
+        });
+    }
+
+    private get needsToVerifyHashCommitment(): boolean {
+        if (this.channel.initiatedByUs) {
```

Comment:
```suggestion
        return this.channel.initiatedByUs;
```
or even inline this?